src/fparser/Fortran2003.py

```diff
@@ -268,38 +287,87 @@ class BlockBase(Base):
 ----------
 content : tuple
     """
+    @staticmethod
     def match(startcls, subclasses, endcls, reader,
-              match_labels = False,
-              match_names = False, set_unspecified_end_name = False,
-              match_name_classes = (),
-              enable_do_label_construct_hook = False,
-              enable_if_construct_hook = False,
-              enable_where_construct_hook = False,
-              enable_select_type_construct_hook = False,
-              enable_case_construct_hook = False
-              ):
+              match_labels=False,
+              match_names=False,
+              set_unspecified_end_name=False,
+              match_name_classes=(),
+              enable_do_label_construct_hook=False,
+              enable_if_construct_hook=False,
+              enable_where_construct_hook=False,
+              enable_select_type_construct_hook=False,
+              enable_case_construct_hook=False):
+        '''
+        Checks whether the content in reader matches the given
+        type of block statement (e.g. DO..END DO, IF...END IF etc.)
+
+        :param type startcls: The class marking the beginning of the block
+        :param list subclasses: List of classes that can be children of
+                                the block
+        :param type endcls: The class marking the end of the block
+        :param reader: Content to check for match
+        :type reader: str or instance of :py:class:`FortranReaderBase`
+        :param bool match_labels: TBD
+        :param bool match_names: TBD
+        :param bool set_unspecified_end_name: TBD
+        :param tuple match_name_classes: TBD
+        :param bool enable_do_label_construct_hook: TBD
+        :param bool enable_if_construct_hook: TBD
+        :param bool enable_where_construct_hook: TBD
+        :param bool enable_select_type_construct_hook: TBD
+        :param bool enable_case_construct_hook: TBD
+
+        :return: instance of startcls or None if no match is found
+        :rtype: startcls
+        '''
         assert isinstance(reader, FortranReaderBase), repr(reader)
         content = []
+
         if startcls is not None:
+            # Deal with any preceding comments
```

Comment:
What about comments after the block? I can see that these might be picked up by the next block but what about the last block in a program. Is it guaranteed that we never get into a case where comments are not dealt with?

Reply:
I've added to an existing test and comments after an 'end program' at least are picked up. However, as you comment below, we really need tests for this for all types of block.